File: WEEK7/Js/bj2_14501.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const n = Number(input[0]);
const t = [];
const p = [];

for (let i = 1; i <= n; i++) {
    const [time, pay] = input[i].split(' ').map(Number);
    t.push(time);
    p.push(pay);
}

const dp = new Int32Array(n + 1);

for (let i = n - 1; i >= 0; i--) {
    if (i + t[i] > n) {
        dp[i] = dp[i + 1];
    } else {
        dp[i] = Math.max(dp[i + 1], p[i] + dp[i + t[i]]);
    }
}

console.log(dp[0]);

File: WEEK7/Python/bj2_14501.py
import sys

n = int(sys.stdin.readline())
t = []
p = []
dp = [0] * (n + 1)

for _ in range(n):
    time, pay = map(int, sys.stdin.readline().split())
    t.append(time)
    p.append(pay)

for i in range(n - 1, -1, -1):
    if i + t[i] > n:
        dp[i] = dp[i + 1]
    else:
        dp[i] = max(dp[i + 1], p[i] + dp[i + t[i]])

print(dp[0])
